File: Java/Quicksort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
public class QuickSort {


    // Function to measure memory usage returns usage in kb
    static long getMemoryUsage() {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        return (runtime.totalMemory() - runtime.freeMemory());/ 1024;
    }

    static void swap(ArrayList<Integer> arr, int x, int y) {
        int temp = arr.get(x);
        arr.set(x, arr.get(y));
        arr.set(y, temp);
    }

    static void quickSort(ArrayList<Integer> arr, int low, int high) {
        if (low < high) {
            int partition = partition(arr, low, high);
            quickSort(arr, low, partition - 1);
            quickSort(arr, partition + 1, high);
        }
    }

    static int partition(ArrayList<Integer> arr, int low, int high) {
        int pivot = arr.get(high);
        int i = low - 1;
        for (int j = low; j <= high - 1; j++) {
            if (arr.get(j) <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter a size: ");
        int size = scanner.nextInt();


        // Define your static array here
        ArrayList<Integer> array = new ArrayList<>(Arrays.asList());

        // Adding size declaration here similar to the older code 

        Random rand = new Random();
        for (int i = 0; i < size; i++) {
            array.add(rand.nextInt(100));


        }
        long start = System.nanoTime();



        // Function call
        QuickSort t = new QuickSort();
        t.quickSort(array, 0, size - 1);

        long stop = System.nanoTime();
        long duration = stop - start;


      

        // Print statements for the execution time, CPU time, and memory
        System.out.println("Quick sort time is: " + (duration / 1000) + " microseconds");
        
        System.out.println("Memory usage: " + getMemoryUsage() + " KB");

        // Print the sorted array to verify
        for (int j = 0; j < size; j++) {
            System.out.print(" " + array.get(j));
        }
        System.out.println();
    }
}

File: C++ Code/For mac/quickSort.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <sys/types.h>
#include <sys/sysctl.h>

using namespace std::chrono;
using namespace std;

// Function to measure memory usage returns usage in kilobytes
long getMemoryUsage() {
    struct rusage rusage;
    getrusage(RUSAGE_SELF, &rusage);
    return rusage.ru_maxrss;
}

// Function to measure CPU usage (macOS specific)
double getCPUUsage() {
    size_t size = sizeof(int);
    int cpuPercent;
    sysctlbyname("kern.cp_time", &cpuPercent, &size, NULL, 0);
    return cpuPercent / 100.0;
}

void swap(int &x, int &y)
{
    int temp;
    temp = x;
    x = y;
    y = temp;
}

void quickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int partition = 0;
        int pivot = arr[high];
        int temp = (low - 1);
        for (int i = low; i <= high - 1; i++)
        {
            if (arr[i] <= pivot)
            {
                temp++;
                swap(arr[temp], arr[i]);
            }
        }
        swap(arr[temp + 1], arr[high]);
        partition = (temp + 1);

        quickSort(arr, low, partition - 1);
        quickSort(arr, partition + 1, high);
    }
}

int main() 
{
    // get random sample size
    vector<int> array;
    int size;
    cout << "Enter a size: ";
    cin >> size;
    
    for (int i = 0; i < size; i++)
    {
        array.push_back(rand() % 100);
        array.push_back(array[i]);
    }

    // Getting the time of competition for the execution, CPU, and memory usage
    auto start = high_resolution_clock::now();
   
    quickSort(array, 0, size - 1);
    
    auto stop = high_resolution_clock::now();
  
    auto duration = duration_cast<microseconds>(stop - start);

    // prints the statements out
    cout << "Insertion sort time is: " << duration.count() << " microseconds" << endl;
    cout << "CPU usage: " << getCPUUsage() * 100 << "%" << endl; // CPU usage in percentage
    cout << "Memory usage: " << getMemoryUsage() << " KB" << endl;

    // prints the list to make sure the algorithm sorts the list properly
    for (int j = 0; j < size; j++) {
        cout << ' ' << array[j];
    }

    cout << endl;
    return 0;
}

